File: gamepack/Decompiled/UserList.java
import java.util.Arrays;
import java.util.Comparator;
import java.util.HashMap;

public abstract class UserList {
   int size0 = 0;
   Comparator comparator = null;
   HashMap usernamesMap;
   HashMap previousUsernamesMap;
   User[] array;
   final int capacity;

   UserList(int var1) {
      this.capacity = var1 * 450790185;
      this.array = this.newTypedArray(var1);
      this.usernamesMap = new HashMap(var1 / 8);
      this.previousUsernamesMap = new HashMap(var1 / 8);
   }

   public void clear() {
      this.size0 = 0;
      Arrays.fill(this.array, (Object)null);
      this.usernamesMap.clear();
      this.previousUsernamesMap.clear();
   }

   public int size() {
      return this.size0 * -940640785;
   }

   public boolean isFull() {
      return -734344935 * this.capacity == -940640785 * this.size0;
   }

   public boolean contains(Username var1) {
      if (!var1.hasCleanName()) {
         return false;
      } else {
         return this.usernamesMap.containsKey(var1) ? true : this.previousUsernamesMap.containsKey(var1);
      }
   }

   public User getByUsername(Username var1) {
      User var3 = this.getByCurrentUsername(var1);
      return var3 != null ? var3 : this.getByPreviousUsername(var1);
   }

   User getByCurrentUsername(Username var1) {
      return !var1.hasCleanName() ? null : (User)this.usernamesMap.get(var1);
   }

   User getByPreviousUsername(Username var1) {
      return !var1.hasCleanName() ? null : (User)this.previousUsernamesMap.get(var1);
   }

   public final boolean removeByUsername(Username var1) {
      User var3 = this.getByCurrentUsername(var1);
      if (null == var3) {
         return false;
      } else {
         this.remove(var3);
         return true;
      }
   }

   final void remove(User var1) {
      int var3 = this.indexOf(var1);
      if (-1 != var3) {
         this.arrayRemove(var3);
         this.mapRemove(var1);
      }
   }

   User addLastNoPreviousUsername(Username var1) {
      return this.addLast(var1, (Username)null);
   }

   User addLast(Username var1, Username var2) {
      if (this.getByCurrentUsername(var1) != null) {
         throw new IllegalStateException();
      } else {
         User var4 = this.newInstance();
         var4.set(var1, var2);
         this.arrayAddLast(var4);
         this.mapPut(var4);
         return var4;
      }
   }

   public final User get(int var1) {
      if (var1 >= 0 && var1 < this.size0 * -940640785) {
         return this.array[var1];
      } else {
         throw new ArrayIndexOutOfBoundsException(var1);
      }
   }

   public final void sort() {
      if (null == this.comparator) {
         Arrays.sort(this.array, 0, -940640785 * this.size0);
      } else {
         Arrays.sort(this.array, 0, -940640785 * this.size0, this.comparator);
      }

   }

   final void changeName(User var1, Username var2, Username var3) {
      this.mapRemove(var1);
      var1.set(var2, var3);
      this.mapPut(var1);
   }

   final int indexOf(User var1) {
      for(int var3 = 0; var3 < -940640785 * this.size0; ++var3) {
         if (this.array[var3] == var1) {
            return var3;
         }
      }

      return -1;
   }

   final void mapRemove(User var1) {
      if (this.usernamesMap.remove(var1.username0) == null) {
         throw new IllegalStateException();
      } else {
         if (var1.previousUsername != null) {
            this.previousUsernamesMap.remove(var1.previousUsername);
         }

      }
   }

   final void arrayAddLast(User var1) {
      this.array[(this.size0 += 905500943) * -940640785 - 1] = var1;
   }

   final void mapPut(User var1) {
      this.usernamesMap.put(var1.username0, var1);
      if (var1.previousUsername != null) {
         User var3 = (User)this.previousUsernamesMap.put(var1.previousUsername, var1);
         if (null != var3 && var3 != var1) {
            var3.previousUsername = null;
         }
      }

   }

   final void arrayRemove(int var1) {
      this.size0 -= 905500943;
      if (var1 < -940640785 * this.size0) {
         System.arraycopy(this.array, 1 + var1, this.array, var1, -940640785 * this.size0 - var1);
      }

   }

   public final void removeComparator() {
      this.comparator = null;
   }

   public final void addComparator(Comparator var1) {
      if (this.comparator == null) {
         this.comparator = var1;
      } else if (this.comparator instanceof AbstractUserComparator) {
         ((AbstractUserComparator)this.comparator).ay(var1);
      }

   }

   abstract User[] newTypedArray(int var1);

   abstract User newInstance();
}

File: find-names/src/main/java/mapper/identifiers/classes/UserList.kt
package mapper.identifiers.classes

import mapper.abstractclasses.IdentityMapper
import mapper.abstractclasses.OrderMapper
import mapper.annotations.DependsOn
import mapper.annotations.MethodParameters
import mapper.predicateutilities.*
import mapper.wrappers.ClassWrapper
import mapper.wrappers.FieldWrapper
import mapper.wrappers.InstructionWrapper
import mapper.wrappers.MethodWrapper
import org.objectweb.asm.Opcodes.*
import org.objectweb.asm.Type.*
import org.runestar.client.common.startsWith
import java.lang.reflect.Modifier

@DependsOn(User::class)
class UserList : IdentityMapper.Class() {

    override val predicate = predicateOf<ClassWrapper> { it.superType == Any::class.type }
            .and { it.instanceFields.any { it.type == type<User>().withDimensions(1) } }

    @MethodParameters()
    class newInstance : IdentityMapper.InstanceMethod() {
        override val predicate = predicateOf<MethodWrapper> { Modifier.isAbstract(it.access) }
                .and { it.returnType == type<User>() }
    }

    @MethodParameters("size")
    class newTypedArray : IdentityMapper.InstanceMethod() {
        override val predicate = predicateOf<MethodWrapper> { Modifier.isAbstract(it.access) }
                .and { it.returnType == type<User>().withDimensions(1) }
    }

    class array : IdentityMapper.InstanceField() {
        override val predicate = predicateOf<FieldWrapper> { it.type == type<User>().withDimensions(1) }
    }

    class comparator : IdentityMapper.InstanceField() {
        override val predicate = predicateOf<FieldWrapper> { it.type == Comparator::class.type }
    }

    @MethodParameters()
    class clear : IdentityMapper.InstanceMethod() {
        override val predicate = predicateOf<MethodWrapper> { it.instructions.any { it.isMethod && it.methodId == HashMap<*, *>::clear.id } }
    }

    class size0 : OrderMapper.InConstructor.Field(UserList::class, 0) {
        override val predicate = predicateOf<InstructionWrapper> { it.opcode == PUTFIELD && it.fieldType == INT_TYPE }
    }

    class capacity : OrderMapper.InConstructor.Field(UserList::class, 1) {
        override val predicate = predicateOf<InstructionWrapper> { it.opcode == PUTFIELD && it.fieldType == INT_TYPE }
    }

    @MethodParameters("index")
    class get : IdentityMapper.InstanceMethod() {
        override val predicate = predicateOf<MethodWrapper> { it.returnType == type<User>() }
                .and { it.instructions.any { it.opcode == NEW && it.typeType == ArrayIndexOutOfBoundsException::class.type } }
    }

    @MethodParameters()
    class sort : IdentityMapper.InstanceMethod() {
        override val predicate = predicateOf<MethodWrapper> { it.returnType == VOID_TYPE && it.instructions.any { it.isMethod && it.methodName == "sort" } }
    }

    class usernamesMap : OrderMapper.InConstructor.Field(UserList::class, 0) {
        override val predicate = predicateOf<InstructionWrapper> { it.opcode == PUTFIELD && it.fieldType == HashMap::class.type }
    }

    class previousUsernamesMap : OrderMapper.InConstructor.Field(UserList::class, 1) {
        override val predicate = predicateOf<InstructionWrapper> { it.opcode == PUTFIELD && it.fieldType == HashMap::class.type }
    }

    @MethodParameters("user")
    class mapRemove : IdentityMapper.InstanceMethod() {
        override val predicate = predicateOf<MethodWrapper> { it.returnType == VOID_TYPE }
                .and { it.arguments.size in 1..2 }
                .and { it.arguments.startsWith(type<User>()) }
                .and { it.instructions.any { it.isMethod && it.methodName == "remove" } }
    }

    @MethodParameters("user")
    class mapPut : IdentityMapper.InstanceMethod() {
        override val predicate = predicateOf<MethodWrapper> { it.returnType == VOID_TYPE }
                .and { it.arguments.size in 1..2 }
                .and { it.arguments.startsWith(type<User>()) }
                .and { it.instructions.any { it.isMethod && it.methodName == "put" } }
    }

    @MethodParameters("username")
    class contains : IdentityMapper.InstanceMethod() {
        override val predicate = predicateOf<MethodWrapper> { it.returnType == BOOLEAN_TYPE }
                .and { it.arguments.size in 1..2 }
                .and { it.arguments.startsWith(type<Username>()) }
                .and { it.instructions.any { it.isMethod && it.methodName == "containsKey" } }
    }

    @MethodParameters()
    @DependsOn(capacity::class)
    class isFull : IdentityMapper.InstanceMethod() {
        override val predicate = predicateOf<MethodWrapper> { it.returnType == BOOLEAN_TYPE }
                .and { it.arguments.size in 0..1 }
                .and { it.instructions.any { it.isField && it.fieldId == field<capacity>().id } }
    }

    @MethodParameters()
    class size : IdentityMapper.InstanceMethod() {
        override val predicate = predicateOf<MethodWrapper> { it.returnType == INT_TYPE }
                .and { it.arguments.size in 0..1 }
                .and { it.instructions.none { it.opcode == IINC } }
    }

    @MethodParameters("user", "username", "previousUsername")
    class changeName : IdentityMapper.InstanceMethod() {
        override val predicate = predicateOf<MethodWrapper> { it.returnType == VOID_TYPE }
                .and { it.arguments.size in 3..4 }
                .and { it.arguments.startsWith(type<User>(), type<Username>(), type<Username>()) }
    }

    @MethodParameters("user")
    class indexOf : IdentityMapper.InstanceMethod() {
        override val predicate = predicateOf<MethodWrapper> { it.returnType == INT_TYPE }
                .and { it.arguments.size in 1..2 }
                .and { it.arguments.startsWith(type<User>()) }
                .and { it.instructions.any { it.opcode == ICONST_M1 } }
    }

    @MethodParameters("username")
    @DependsOn(usernamesMap::class)
    class getByCurrentUsername : IdentityMapper.InstanceMethod() {
        override val predicate = predicateOf<MethodWrapper> { it.returnType == type<User>() }
                .and { it.arguments.size in 1..2 }
                .and { it.arguments.startsWith(type<Username>()) }
                .and { it.instructions.any { it.opcode == GETFIELD && it.fieldId == field<usernamesMap>().id } }
    }

    @MethodParameters("previousUsername")
    @DependsOn(previousUsernamesMap::class)
    class getByPreviousUsername : IdentityMapper.InstanceMethod() {
        override val predicate = predicateOf<MethodWrapper> { it.returnType == type<User>() }
                .and { it.arguments.size in 1..2 }
                .and { it.arguments.startsWith(type<Username>()) }
                .and { it.instructions.any { it.opcode == GETFIELD && it.fieldId == field<previousUsernamesMap>().id } }
    }

    @MethodParameters("username")
    @DependsOn(getByPreviousUsername::class)
    class getByUsername : IdentityMapper.InstanceMethod() {
        override val predicate = predicateOf<MethodWrapper> { it.returnType == type<User>() }
                .and { it.arguments.size in 1..2 }
                .and { it.arguments.startsWith(type<Username>()) }
                .and { it.instructions.any { it.isMethod && it.methodId == method<getByPreviousUsername>().id } }
    }

    @MethodParameters("user")
    class arrayAddLast : IdentityMapper.InstanceMethod() {
        override val predicate = predicateOf<MethodWrapper> { it.returnType == VOID_TYPE }
                .and { it.arguments.size in 1..2 }
                .and { it.arguments.startsWith(type<User>()) }
                .and { it.instructions.none { it.isMethod } }
    }

    @MethodParameters("username", "previousUsername")
    class addLast : IdentityMapper.InstanceMethod() {
        override val predicate = predicateOf<MethodWrapper> { it.returnType == type<User>() }
                .and { it.arguments.size in 2..3 }
                .and { it.instructions.any { it.opcode == NEW && it.typeType == IllegalStateException::class.type } }
    }

    @MethodParameters("username")
    @DependsOn(addLast::class)
    class addLastNoPreviousUsername : IdentityMapper.InstanceMethod() {
        override val predicate = predicateOf<MethodWrapper> { it.returnType == type<User>() }
                .and { it.arguments.size in 1..2 }
                .and { it.instructions.any { it.isMethod && it.methodId == method<addLast>().id } }
    }

    @MethodParameters("index")
    class arrayRemove : IdentityMapper.InstanceMethod() {
        override val predicate = predicateOf<MethodWrapper> { it.returnType == VOID_TYPE }
                .and { it.arguments.size in 1..2 }
                .and { it.instructions.any { it.isMethod && it.methodName == "arraycopy" } }
    }

    @MethodParameters("user")
    @DependsOn(arrayRemove::class)
    class remove : IdentityMapper.InstanceMethod() {
        override val predicate = predicateOf<MethodWrapper> { it.returnType == VOID_TYPE }
                .and { it.arguments.size in 1..2 }
                .and { it.instructions.any { it.isMethod && it.methodId == method<arrayRemove>().id } }
    }

    @MethodParameters("username")
    @DependsOn(remove::class)
    class removeByUsername : IdentityMapper.InstanceMethod() {
        override val predicate = predicateOf<MethodWrapper> { it.returnType == BOOLEAN_TYPE }
                .and { it.arguments.size in 1..2 }
                .and { it.instructions.any { it.isMethod && it.methodId == method<remove>().id } }
    }

    @MethodParameters()
    @DependsOn(comparator::class)
    class removeComparator : IdentityMapper.InstanceMethod() {
        override val predicate = predicateOf<MethodWrapper> { !it.arguments.startsWith(Comparator::class.type) }
                .and { it.instructions.any { it.opcode == PUTFIELD && it.fieldId == field<comparator>().id } }
    }

    @MethodParameters("c")
    @DependsOn(comparator::class)
    class addComparator : IdentityMapper.InstanceMethod() {
        override val predicate = predicateOf<MethodWrapper> { it.arguments.startsWith(Comparator::class.type) }
    }
}